client/galaxy/scripts/components/ToolRecommendation.vue

```diff
@@ -0,0 +1,282 @@
+                            }).responseText
+                        ),
+                        extToType = datatypes_mapping.ext_to_class_name,
+                        typeToType = datatypes_mapping.class_to_classes,
+                        predData = data.predicted_data;
+                    this.deprecated = predData["is_deprecated"];
+                    if (data !== null && predData.children.length > 0) {
+                        const filteredData = {},
+                            compatibleTools = {},
+                            filteredChildren = [],
+                            outputDatatypes = predData["o_extensions"],
+                            children = predData.children;
+                        for (const nameObj of children.entries()) {
+                            const inputDatatypes = nameObj[1]["i_extensions"];
+                            for (const out_t of outputDatatypes.entries()) {
+                                for (const in_t of inputDatatypes.entries()) {
+                                    const child = extToType[out_t[1]],
+                                        parent = extToType[in_t[1]];
+                                    if (
+                                        (typeToType[child] && parent in typeToType[child]) === true ||
+                                        out_t[1] === "input" ||
+                                        out_t[1] === "_sniff_" ||
+                                        out_t[1] === "input_collection"
+                                    ) {
+                                        compatibleTools[nameObj[1]["tool_id"]] = nameObj[1]["name"];
+                                        break;
+                                    }
+                                }
+                            }
+                        }
+                        for (const id in compatibleTools) {
+                            for (const nameObj of children.entries()) {
+                                if (nameObj[1]["tool_id"] === id) {
+                                    filteredChildren.push(nameObj[1]);
+                                    break;
+                                }
+                            }
+                        }
+                        filteredData["o_extensions"] = predData["o_extensions"];
+                        filteredData["name"] = predData["name"];
+                        filteredData["children"] = filteredChildren;
+                        if (filteredChildren.length > 0 && this.deprecated === false) {
+                            self.renderD3Tree(filteredData);
+                        } else if (this.deprecated === true) {
+                            this.deprecatedMessage = predData["message"];
+                        }
+                    }
+                }
+            });
+        },
+        renderD3Tree: function(predictedTools) {
+            const margin = { top: 20, right: 30, bottom: 20, left: 250 },
```

Comment:
I am a little worried about this margins, if these constants are needed please specifiy them explicitly at the top of the file as constants.